### PR TITLE
pydoc-markdown + google style docstrings

### PR DESCRIPTION
**Pull request by eode**
## Description
Options we had:
* sphinx
  * Reads Google docstrings
  * More fomalized API
  * Very rudimentary (read: mostly not done) markdown output
  * ended up that we'd be mixing Google Docstrings + RST to get the (hopefully) right markdown output
* pydoc-markdown
  * rough
  * less-formalized API
  * No Google docstrings support
  * Ends up we'd be mixing Google Docstrings + Markdown to get Markdown -- not bad.

in the end I used `pydoc-markdown`:
 * Most changes made in `google-docstrings.py`
 * We'll probably find areas our docstrings don't match the format exactly.
 * **The modified copy of `pydoc-markdown` is ~in-tree under `/gendocs`~ in a quilt repository [here](https://github.com/quiltdata/pydoc-markdown) **
 * There is some conformity to `pydoc-markdown`'s style and structure needed, but not too much.
  * Basically, Google Docstrings with markdown support.
    * ..and don't use section headers inside your docstrings.
    * ..but formatting, lists, and code blocks for various languages should pass through nicely.
    * formatting defaults to Python.

### CODE REVIEW

**Comment by eode** [comment]
This may need more docstrings updated, but I wanted to get this out here for review.

```bash
cd <git root>/gendocs
pip install --editable ./pydoc-markdown   # once
pydocmd generate  # whenever you want to update docs

# edit config
nano pydocmd.yml

# built files
cat _build/pydocmd/api_Bucket.md
cat _build/pydocmd/api_Package.md
cat _build/pydocmd/api.md
```

I don't imagine those are the exact filenames we'll want, and very likely isn't the final doc structure we want, but it should suffice for examples.

Github-readable examples are in the PR as well under `gendocs/example_generated`.  I can remove them from the PR just prior to merging, they're just here to view.

**Comment by ResidentMario** [comment]
We should open our own fork of `pydoc-markdown` that's hosted on the `quiltdata` GitHub organization, and set up instructions telling the user to draw from there.

**Comment by eode** [comment]
@ResidentMario Ok, see if this works nicely for you.

```bash
# from gendocs dir:
# If in your t4 virtualenv, do this:
 ./build.py
# otherwise, do
/path/to/t4/virtualenv/bin/python3 build.py
```

..note that this (politely) replaces the version of pydocmd that you have installed, if it's not the expected version.

**Comment by eode** [comment]
> We should open our own fork of `pydoc-markdown` that's hosted on the `quiltdata` GitHub organization, and set up instructions telling the user to draw from there.

Done, and changes to accommodate that are done.

**Comment by ResidentMario** [comment]
There are some minor issues with e.g. the capitalization of things which don't matter. It'd be easiest to merge this PR, then do another PR lightly correcting any docstrings with residual issues.

Can you add `api.py` to the generated documentation files as well? There are top-level methods in there that are also important to the user.

Can you move the script output target to `t4/docs/API Reference`?

**Comment by eode** [comment]
Note:
Heading IDs are a bit ugly right now.  However, GitBooks has a nice workaround for this using their own flavor of markdown.  This is easy to implement/fix to make the heading IDs a lot nicer, but is probably better done in another PR.
